stablehlo/dialect/TypeInference.cpp

```diff
@@ -3869,6 +3996,59 @@ LogicalResult verifyReduceWindowOp(
   return success();
 }
 
+LogicalResult verifyReshapeOpQuantizationConstraints(
+    std::optional<Location> location, Type operandTy, Type resultTy) {
+  // reshape_c1
+  if (allQuantized<quant::UniformQuantizedType>(operandTy, resultTy))
+    if (getElementTypeOrSelf(operandTy) != getElementTypeOrSelf(resultTy))
+      return emitOptionalError(location,
+                               "expect same quantization scale and zero_point "
+                               "for operand and result but got ",
+                               operandTy, " vs ", resultTy);
+
+  // reshape_c1
+  if (allQuantized<quant::UniformQuantizedPerAxisType>(operandTy, resultTy)) {
+    if (!isSameQuantPerAxisScaleZeroPoint(operandTy, resultTy)) {
+      return emitOptionalError(
+          location,
+          "expect same quantization scales and zero_points "
+          "for operand and result but got ",
+          operandTy, " vs ", resultTy);
+    }
+    // reshape_c3
+    auto operandQDim = getElementTypeOrSelf(operandTy)
+                           .cast<quant::UniformQuantizedPerAxisType>()
+                           .getQuantizedDimension();
+    auto resultQDim = getElementTypeOrSelf(resultTy)
+                          .cast<quant::UniformQuantizedPerAxisType>()
+                          .getQuantizedDimension();
+    auto operandShape = operandTy.cast<ShapedType>().getShape();
+    auto resultShape = resultTy.cast<ShapedType>().getShape();
+
+    if (operandTy.cast<ShapedType>().getDimSize(operandQDim) !=
+        resultTy.cast<ShapedType>().getDimSize(resultQDim))
+      return emitOptionalError(
+          location,
+          "expect same quantization dimension size for operand and result ",
+          operandTy, " and ", resultTy);
+
+    uint64_t operandProd = 1;
```

Comment:
The second half of the check is not implemented where we need to compare the product of the dim-sizes past the quantization dimension.  
```
reduce(dims(operand, [quantization_dimension(operand) + 1, ..., rank(operand) - 1]), init_values=1, dimensions=[0], body=lambda x, y: x * y) = reduce(dims(result, [quantization_dimension(result) + 1, ..., rank(result) - 1]), init_values=1, dimensions=[0], body=lambda x, y: x * y).
```

Reply:
```
    isCompatibleQuantizationDimention = [&](operandShape, resultShape, int64_t begin_pos, int64_end_pos) {
       uint64_t operandProd = 1;
       std::for_each(operandShape.begin() + begin_pos, operandShape.begin() + end_pos,
                  [&operandProd](int32_t dimSize) { operandProd *= dimSize; });
        uint64_t resultProd = 1;
        std::for_each(resultShape.begin() + negin_pos, resultShape.begin() + end_pos,
                  [&resultProd](int32_t dimSize) { resultProd *= dimSize; });
     };
    

   return isCompatibleQuantizationDimention(operandShape, resultShape, 0, operandQDim) && isCompatibleQuantizationDimention(operandShape, resultShape,  operandQDim, resultShape.size()-1)
```

Reply:
I added check for the second half, however it is already covered by other checks and I was unable to simulate test condition. 

1. there is a check which validate operand element count == result element count
2. product of dims before quantization dim are equal for operand and result (first half)
3. quantization dim sizes are equal. 

after these checks, 2nd half of the check is already verified. We can add placeholder code but can't add test for it. So decided to drop the changes for second half of the constraint.

Reply:
Thanks for explaining. This makes sense!